types/index.d.ts

```diff
@@ -31,68 +35,112 @@ interface AssembleOptions {
 export function discoverProtocol(transport: Service.Transport, callback: Callback<any>): void;
 export function extractFileHeader(filePath: string, options?: any): void;
-export function parse(schemaOrProtocolIdl: string, options?: any): Service.Protocol | Type.Type; // TODO
-export function readProtocol(protocolIdl: string, options?: Partial<ReaderOptions>): Service.Protocol;
+export function infer(args: any): any;
+export function parse(schemaOrProtocolIdl: string, options?: any): Protocol | Type; // TODO
+export function readProtocol(protocolIdl: string, options?: Partial<ReaderOptions>): Protocol;
 export function readSchema(schemaIdl: string, options?: Partial<ReaderOptions>): Schema;
 // TODO streams
-// TODO types
-
-export namespace Type {
-  interface Type {
-    // TODO clone(val, opts)
-    // TODO compare
-    // TODO compareBuffers(buf1, buf2)
-    // TODO createResolver(type, opts)
-    // TODO decode(buf, pos, resolver)
-    // TODO encode(val, buf, pos)
-    // TODO equals(type)
-    // TODO fingerprint(algorithm)
-    fromBuffer(buffer: Buffer, resolver: any, noCheck: boolean): Type; // TODO
-    // TODO fromString(str)
-    // TODO inspect()
-    // TODO isValid(val, opts)
-    // TODO random()
-    // TODO schema(opts)
-    toBuffer(value: object): Buffer;
-    // TODO toJSON()
-    // TODO toString(val)
-    // TODO wrap(val)
-  }
-
-  function forSchema(schema: Schema): Type;
-  function forValue(value: object): Type;
-  // TODO function forTypes(types, opts)
-  function isType(arg: any): boolean; // TODO
+
+// TODO more specific types than `any`
+export class Type {
+  clone(val: any, opts?: any): any;
+  compare(val1: any, val2: any): number;
+  compareBuffers(buf1: any, buf2: any): number;
+  constructor(schema: Schema, opts?: any);
+  createResolver(type: any, opts?: any): any;  // TODO: opts not documented on wiki
+  decode(buf: any, pos?: any, resolver?: any): any
+  fingerprint(algorithm?: any): any;
+  fromBuffer(buffer: Buffer, resolver?: any, noCheck?: boolean): Type; // TODO
+  fromString(str: any): any;
+  inspect(): string;
+  isValid(val: any, opts?: any): any;
+  random(): Type;
+  schema(opts?: any): any;
+  toBuffer(value: object): Buffer;
+  toJSON(): string;
+  toString(val?: any): any;
+  wrap(val: any): any;
+  readonly aliases: string[]|undefined;
+  readonly doc: string|undefined;
+  readonly name: string|undefined;
+  readonly branchName: string|undefined;
+  readonly typeName: string;
+  static forSchema(schema: Schema, opts?: any): Type;
+  static forTypes(types: any, opts?: any): Type;
+  static forValue(value: object, opts?: any): Type;
+  static isType(arg: any, ...prefix: string[]): boolean;
+  static __reset(size: number): void;
 }
 
-export namespace Service {
-  interface Protocol {
-    [key: string]: any; // TODO object should be further specified
-  }
+// Deprecated, but kept for now because this is referenced elsewhere (e.g. Service.forProtocol())
```

Comment:
Actually, all references to `Protocol` in this file should be `Object` (or `any`, not sure how TypeScript treats vanilla objects). For example `readProtocol`'s return value doesn't have any of the methods below. If we wanted to be more accurate, we could define a custom interface with the expected attributes in the object (e.g. `"messages"`, `"protocol"`, and `"types"`) but it's probably fine to do without for now.